networks.py

```diff
@@ -268,6 +269,39 @@ def step(self, inputs, prev_state):
   def unroll(self, inputs, prev_state):
     return inputs[1], ()
 
+class MultiHeadAttentionBlock(snt.Module):
+  def __init__(self, num_heads, output_size, name='AttentionBlock'):
+    super(MultiHeadAttentionBlock, self).__init__()
+    self.num_heads = num_heads
+    self.W_K = []
+    self.W_V = []
+    self.W_Q = []
+    for _ in range(num_heads):
+      self.W_K.append(snt.Linear(int(output_size/num_heads))) #output is d_model/num_heads
+      self.W_V.append(snt.Linear(int(output_size/num_heads))) #output is d_model/num_heads
+      self.W_Q.append(snt.Linear(int(output_size/num_heads))) #output is d_model/num_heads
+    self.W_O = snt.Linear(output_size)
+
+  def initial_state(self, batch_size):
+    raise NotImplementedError()
+  
+  def attention(self, queries, keys, values):
+    compat = queries * tf.transpose(keys) # check rank
+    norm_compat = compat / (keys.shape[-1] ** .05) # check shape
+    probs = tf.nn.softmax(norm_compat)
+    att = probs * values
+    return att
+
+  def __call__(self, queries, keys, values):
```

Comment:
Call should take `inputs: [B, S, D_M]`

(assuming only 1 head)

queries = self.W_Q(inputs): [B, S, D_K]
keys = self.W_K(keys): [B, S, D_K]
values = self.W_V(values): [B, S, D_V]

may need snt.BatchApply(self.W_Q) instead of regular self.W_Q

return attention(queries, keys, values): [B, S, D_V]